File: src/MyServiceBus.PostgreSql/PostgreSqlScheduledWorkSource.cs
using Npgsql;
using NpgsqlTypes;

namespace MyServiceBus.Persistence.PostgreSql;

/// <summary>
/// Reads scheduled message state from the PostgreSQL outbox without loading message bodies.
/// </summary>
public sealed class PostgreSqlScheduledWorkSource : IScheduledWorkSource
{
    private readonly NpgsqlDataSource dataSource;
    private readonly string serviceName;

    public PostgreSqlScheduledWorkSource(NpgsqlDataSource dataSource, string serviceName)
    {
        this.dataSource = dataSource ?? throw new ArgumentNullException(nameof(dataSource));
        ArgumentException.ThrowIfNullOrWhiteSpace(serviceName);
        this.serviceName = serviceName;
    }

    public string Provider => "PostgreSQL";

    public bool Authoritative => true;

    public async Task<IReadOnlyList<ScheduledWorkState>> GetSnapshotAsync(
        int maximumCount,
        CancellationToken cancellationToken = default)
    {
        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(maximumCount);
        const string sql = """
            SELECT message_id, intent, destination_address, message_types[1], scheduled_at_utc,
                state, attempt_count,
                CASE state
                    WHEN 2 THEN COALESCE(dispatched_at_utc, created_at_utc)
                    WHEN 4 THEN COALESCE(cancelled_at_utc, created_at_utc)
                    ELSE created_at_utc
                END AS updated_at_utc,
                failure_category
            FROM myservicebus.outbox_message
            WHERE service_name = @service_name AND scheduled_at_utc IS NOT NULL
            ORDER BY
                CASE WHEN state IN (0, 1) THEN 0 ELSE 1 END,
                CASE WHEN state IN (0, 1) THEN scheduled_at_utc END ASC NULLS LAST,
                CASE WHEN state NOT IN (0, 1) THEN
                    COALESCE(cancelled_at_utc, dispatched_at_utc, created_at_utc)
                END DESC NULLS LAST
            LIMIT @maximum_count;
            """;

        await using var command = dataSource.CreateCommand(sql);
        command.Parameters.AddWithValue("service_name", NpgsqlDbType.Text, serviceName);
        command.Parameters.AddWithValue("maximum_count", NpgsqlDbType.Integer, maximumCount);
        await using var reader = await command.ExecuteReaderAsync(cancellationToken);
        var items = new List<ScheduledWorkState>();
        while (await reader.ReadAsync(cancellationToken))
        {
            var state = reader.GetInt16(5);
            var (status, providerStatus) = MapStatus(state);
            items.Add(new ScheduledWorkState(
                reader.GetGuid(0),
                Provider,
                ScheduleMessageProviderDurability.Durable,
                "Message",
                reader.GetString(3),
                ((Persistence.OutboxDeliveryIntent)reader.GetInt16(1)).ToString(),
                reader.GetString(2),
                reader.GetFieldValue<DateTimeOffset>(4),
                status,
                providerStatus,
                reader.GetInt32(6),
                reader.GetFieldValue<DateTimeOffset>(7),
                reader.IsDBNull(8) ? null : reader.GetString(8)));
        }
        return items;
    }

    private static (ScheduledWorkStatus Status, string ProviderStatus) MapStatus(short state) => state switch
    {
        0 => (ScheduledWorkStatus.Pending, "Pending"),
        1 => (ScheduledWorkStatus.Running, "Leased"),
        2 => (ScheduledWorkStatus.Completed, "Dispatched"),
        3 => (ScheduledWorkStatus.Failed, "Dead"),
        4 => (ScheduledWorkStatus.Cancelled, "Cancelled"),
        _ => throw new InvalidOperationException($"Unknown PostgreSQL outbox state '{state}'.")
    };
}

File: src/Java/myservicebus-postgresql/src/main/java/com/myservicebus/persistence/postgresql/PostgreSqlScheduledWorkSource.java
package com.myservicebus.persistence.postgresql;

import com.myservicebus.SchedulingDurability;
import com.myservicebus.ScheduledWorkSource;
import com.myservicebus.ScheduledWorkState;
import com.myservicebus.ScheduledWorkStatus;
import com.myservicebus.persistence.OutboxDeliveryIntent;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import javax.sql.DataSource;

/** Reads scheduled message state from the PostgreSQL outbox without loading message bodies. */
public final class PostgreSqlScheduledWorkSource implements ScheduledWorkSource {
    private final DataSource dataSource;
    private final String serviceName;

    public PostgreSqlScheduledWorkSource(DataSource dataSource, String serviceName) {
        this.dataSource = Objects.requireNonNull(dataSource, "dataSource");
        if (serviceName == null || serviceName.isBlank()) {
            throw new IllegalArgumentException("serviceName must not be blank");
        }
        this.serviceName = serviceName;
    }

    @Override
    public String getProvider() {
        return "PostgreSQL";
    }

    @Override
    public boolean isAuthoritative() {
        return true;
    }

    @Override
    public CompletionStage<List<ScheduledWorkState>> getSnapshot(int maximumCount) {
        if (maximumCount <= 0) {
            throw new IllegalArgumentException("maximumCount must be greater than zero");
        }
        try {
            return CompletableFuture.completedFuture(query(maximumCount));
        } catch (Exception failure) {
            return CompletableFuture.failedFuture(failure);
        }
    }

    private List<ScheduledWorkState> query(int maximumCount) throws Exception {
        String sql = """
                SELECT message_id, intent, destination_address, message_types[1], scheduled_at_utc,
                    state, attempt_count,
                    CASE state
                        WHEN 2 THEN COALESCE(dispatched_at_utc, created_at_utc)
                        WHEN 4 THEN COALESCE(cancelled_at_utc, created_at_utc)
                        ELSE created_at_utc
                    END AS updated_at_utc,
                    failure_category
                FROM myservicebus.outbox_message
                WHERE service_name = ? AND scheduled_at_utc IS NOT NULL
                ORDER BY
                    CASE WHEN state IN (0, 1) THEN 0 ELSE 1 END,
                    CASE WHEN state IN (0, 1) THEN scheduled_at_utc END ASC NULLS LAST,
                    CASE WHEN state NOT IN (0, 1) THEN
                        COALESCE(cancelled_at_utc, dispatched_at_utc, created_at_utc)
                    END DESC NULLS LAST
                LIMIT ?
                """;
        try (Connection connection = dataSource.getConnection();
                PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setString(1, serviceName);
            statement.setInt(2, maximumCount);
            try (ResultSet result = statement.executeQuery()) {
                List<ScheduledWorkState> items = new ArrayList<>();
                while (result.next()) {
                    short state = result.getShort(6);
                    items.add(new ScheduledWorkState(
                            result.getObject(1, java.util.UUID.class),
                            getProvider(),
                            SchedulingDurability.DURABLE,
                            "Message",
                            result.getString(4),
                            titleCase(OutboxDeliveryIntent.values()[result.getShort(2)].name()),
                            result.getString(3),
                            result.getObject(5, OffsetDateTime.class).toInstant(),
                            mapStatus(state),
                            mapProviderStatus(state),
                            result.getInt(7),
                            result.getObject(8, OffsetDateTime.class).toInstant(),
                            result.getString(9)));
                }
                return List.copyOf(items);
            }
        }
    }

    private static ScheduledWorkStatus mapStatus(short state) {
        return switch (state) {
            case 0 -> ScheduledWorkStatus.PENDING;
            case 1 -> ScheduledWorkStatus.RUNNING;
            case 2 -> ScheduledWorkStatus.COMPLETED;
            case 3 -> ScheduledWorkStatus.FAILED;
            case 4 -> ScheduledWorkStatus.CANCELLED;
            default -> throw new IllegalStateException("Unknown PostgreSQL outbox state '" + state + "'.");
        };
    }

    private static String mapProviderStatus(short state) {
        return switch (state) {
            case 0 -> "Pending";
            case 1 -> "Leased";
            case 2 -> "Dispatched";
            case 3 -> "Dead";
            case 4 -> "Cancelled";
            default -> throw new IllegalStateException("Unknown PostgreSQL outbox state '" + state + "'.");
        };
    }

    private static String titleCase(String value) {
        String lower = value.toLowerCase(java.util.Locale.ROOT);
        return Character.toUpperCase(lower.charAt(0)) + lower.substring(1);
    }
}
